### DISCUSSION TITLE
Add total count field in the response for paginated get clients request

### DISCUSSION THREAD

**Discussion by akoserwal** [keycloak/keycloak#8521] [Archive]
https://issues.redhat.com/browse/KEYCLOAK-17822

Suggestion from Stian
a) Add count in the same way as done for users, without changing the current API
b) Add a new separate client API that can follow the new REST guidelines that are still in design https://github.com/keycloak/keycloak/discussions/8499

**Comment by stianst** [comment]
Those are options, not suggestions from me ;)

**Comment by stianst** [comment]
I fully appreciate that the current Admin APIs are not all that well designed, and pagination is one of the areas that is not done very well. However, we simply can't break the current APIs, and also it is difficult to evolve and improve them.

**Reply by akoserwal** [reply]
A new end-point can be exposed which can support total with clients array in the response. but it goes against the guideline. https://github.com/keycloak/keycloak-community/blob/master/design/rest-api-guideline.md#pagination

Add count in the same way as done for users, without changing the current API:- This solves the purpose but clients would require making 2 calls. First to get total count & second one to get clients with first & max.

**Reply by dgozalo** [reply]
@stianst Would it be possible to consider versioning by content negotiation headers (`application/vnd.myapi.v2+json`) in the Admin APIs? It would solve any future developments.

On the contrary, a different versioning strategy is being used for the Users API, and having two different versioning schemes would be confusing.

**Reply by stianst** [reply]
@dgozalo The current Admin APIs are fairly badly designed, and should be broken into smaller independent APIs. So I don't think the new APIs should be retrofitted onto the existing APIs, but rather just serve as brand new APIs with separate endpoints. I think if we get the rest-guidelines to a ready state, and have an example Client APIs, we will get a lot of contributions to extending them and adding other APIs.

**Reply by stianst** [reply]
> A new end-point can be exposed which can support total with clients array in the response. but it goes against the guideline. https://github.com/keycloak/keycloak-community/blob/master/design/rest-api-guideline.md#pagination

We should and can update the rest-guildelines as they where not finished, and this raises a good point with regards to pagination where the current iteration of the guidelines doesn't allow including a count in the same request.

> 
> Add count in the same way as done for users, without changing the current API:- This solves the purpose but clients would require making 2 calls. First to get total count & second one to get clients with first & max.

Yes, that's the limitation of extending on the current APIs. What we don't want to have is extending the current APIs to become even messier than they currently are, by for example introducing multiple different ways to do pagination.

**Comment by wtrocki** [comment]
@stianst do you have jira/github issues that we can follow for tracking the development process of the new  admin API?

**Reply by stianst** [reply]
It's been more an idea than something we have the capacity to start work on right now, but I've added an Idea proposal here https://github.com/keycloak/keycloak/discussions/8983

**Comment by wtrocki** [comment]


**Reply by stianst** [reply]
Sent you an email about this, as this is not a discussion to have in the Keycloak community project.
